pkg/lib/cockpit-components-upload.jsx

```diff
@@ -0,0 +1,157 @@
+            binary: true,
+            payload: "fsreplace1",
+            path: this.destination,
+            superuser: "try",
+            "send-acks": "frames"
+        });
+        this.channel.addEventListener("control", this.on_control.bind(this));
+    }
+
+    // Private methods
+    on_control(event, message) {
+        if (message?.command === "send-acks") {
+            this.count += message.frames;
+
+            console.assert(this.count <= TOTAL_FRAME_WINDOW, "queue size too big", this.count);
+
+            if (this.resolveHandler) {
+                this.resolveHandler();
+                this.resolveHandler = null;
+            }
+
+            if (this.count === TOTAL_FRAME_WINDOW) {
+                if (this.resolveFlush) {
+                    this.resolveFlush();
+                    this.resolveFlush = null;
+                }
+            }
+        }
+    }
+
+    read_chunk(blob) {
+        return new Promise((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onload = (event) => resolve(event.target.result);
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(blob);
+        });
+    }
+
+    async write(chunk) {
+        console.assert(this.count >= 0, "ack count went negative", this.count);
+        this.count -= 1;
+        this.channel.send(chunk);
+
+        if (this.count > 0) {
+            return Promise.resolve(true);
+        } else {
+            return new Promise((resolve, reject) => {
+                this.resolveHandler = resolve;
+            });
+        }
+    }
+
+    async close() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("close", function(event, message) {
+                if (message.problem) {
+                    reject(message);
+                } else {
+                    resolve(message);
+                }
+            });
+            channel.control({ command: "done" });
+        });
+    }
+
+    // Public methods
+
+    // Flush is waiting on acks in flight
+    flush() {
+        if (this.count === TOTAL_FRAME_WINDOW) {
+            return Promise.resolve(true);
+        } else {
+            return new Promise((resolve, reject) => {
+                this.resolveFlush = resolve;
+            });
+        }
+    }
+
+    async upload(file) {
+        const promise = new Promise((resolve, reject) => {
+            this.channel.addEventListener("close", (event, message) => {
+                if (message.tag) {
+                    resolve(message);
+                } else {
+                    reject(message);
+                }
+            });
+        });
+
+        let chunk_start = 0;
+        let send_chunks = 0;
+
+        while (this.channel.valid && chunk_start <= file.size) {
+            const chunk_next = chunk_start + BLOCK_SIZE;
+            const blob = file.slice(chunk_start, chunk_next);
```

Comment:
This looks pretty reliable in terms of always getting the size you want.... that makes me happy.